### DISCUSSION TITLE
TTS - Mixing datasets for FastPitch + HiFiGAN

### DISCUSSION THREAD

**Discussion by gedefet** [NVIDIA-NeMo/NeMo#3939] [Q&A] [answered]
### Discussed in https://github.com/NVIDIA/NeMo/discussions/3678

<div type='discussions-op-text'>

<sup>Originally posted by **gedefet** February 15, 2022</sup>
Hi All.
<br>
At the end of the FastPitch_Finetuning.ipynb tutorial:

https://github.com/NVIDIA/NeMo/blob/main/tutorials/tts/FastPitch_Finetuning.ipynb
<br>
it suggest two different ways to improve performance, one of which is:
<br>
_Mix new speaker data with old speaker data
We recommend to train fastpitch using both old speaker data (LJSpeech in this notebook) and the new speaker data. In this case, please modify the .json when finetuning fastpitch to include speaker information:_
<br>
but it isn't 100% clear how to do that.
<br>
I found another FastPitch_Finetuning.ipynb here: 

https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/tts/FastPitch_Finetuning.ipynb
<br>
Which is different than the other. 
<br>
It seems that it has two functions defined to do the mixing, but it seems that this colab is abandoned. Nevertheless, I think that has very useful code, and that has tackled this part of mixing datasets.
<br>
Is this true? Or do I need something more that is not there?Why it is abandoned?
<br>
Thanks!</div>

**Comment by godspirit00** [comment]
This is what I created to mix the datasets based on my understanding of the tutorial:
```python
import json
import os

oldjson=input("Path to the manifest JSON of the old speaker: ").strip()
if oldjson=="": exit(0)
newjson=input("Path to the manifest JSON of the new speaker: ").strip()
if newjson=="": exit(0)
num_hours=input("How many hours? (default 5)").strip()
if num_hours=="":
    num_hours=5
else:
    num_hours=float(num_hours)

print(f"Gathering {num_hours} hours of recording from old speaker...")
with open(oldjson,encoding="utf8") as oif:
    oj=oif.readlines()

counter=0
outold=[]
for o in oj:
    if o.strip()=="": continue
    j=json.loads(o)
    if counter<=num_hours*60*60:
        try:
            del j['pred_text']
        except:
            pass
        j['speaker']=0
        outold.append(json.dumps(j))
        counter+=j['duration']
    else:
        break

print(f"Gathering {len(outold)} entries of recording from new speaker...")
with open(newjson,encoding="utf8") as nif:
    nj=nif.readlines()
outnew=[]
while 1:
    for n in nj:
        if len(outnew)>len(outold):
            break
        j=json.loads(n)
        j['speaker']=1
        try:
            del j['pred_text']
        except:
            pass
        outnew.append(json.dumps(j))
    if len(outnew)>len(outold):
        break
    

print("Outputting...")
outdir=os.path.split(newjson)[0]
fn=(os.path.split(newjson)[1]).replace(".json","")+"--with--"+(os.path.split(oldjson)[1]).replace(".json","")+".json"
with open(os.path.join(outdir,fn),"w",encoding="utf8") as out:
    out.write("\n".join(outold))
    out.write("\n")
    out.write("\n".join(outnew))

print("Done!")
```
I'm not sure if it is the best way to do the mixing, but I have tried finetuning fastpitch with it and the result sounds good.
Hope this helps.

**Comment by gedefet** [comment]
@godspirit00 thank you very much.

I made to work the functions ```make_sub_file_list()``` and ```mix_file_list()``` from here https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/tts/FastPitch_Finetuning.ipynb,

that I think essentialy does the same. The first one let you extract a subdataset of n minutes or m audio samples of the complete HiFiTTS.

But, It mixes different speakers from the HiFiTTS dataset. This is, the new dataset.

I think the idea is to mix it with the LJSpeech dataset used in the checkpoint you downloaded for doing the finetuning from that, is that correct? And then do the finetuning not in 5 minutes of your audio but in 5h from the LJSpeech + 30min of your audio? This would be around 330k training steps for FastPitch.

I'm a bit lost there. Maybe @Oktai15 can throw some light here?

BTW, how did you mix it with the LJSpeech dataset? Because it consists only of a list of audios and a .csv.

Thanks!

**Comment by godspirit00** [comment, marked as answer]
@gedefet

> BTW, how did you mix it with the LJSpeech dataset? Because it consists only of a list of audios and a .csv.

See `__process_data()` in `scripts/dataset_processing/tts/ljspeech/get_data.py`.

**Comment by arinaruck** [comment]
Hello! 
Is there any trace of the mentioned `make_sub_file_list()` and `mix_file_list()`?
Currently the colab notebook and the github tutorial are the same

I would also like to revisit the data balance question from @gedefet, is there any info on how to select the amount of "old speaker" data? 
Should the totalling duration be the same or should the amount of the old data me scaled proportionally to the amount of the new data?
I have almost 3 hours of the new speaker, so I am not able to provide 60 times that of the old speaker data
And does the data mixing fine-tuning setup scale for the larger amounts of data or is it beneficial only when very limited data of the new speaker is accessible?

**Reply by gedefet** [reply]
@arinaruck , about the data balance, you should look for the total amount of data. Anyway, you'll have to select the balance that works best for you. No 50 50 is always the best. I suggest to put more data of your speaker.

**Reply by arinaruck** [reply]
Thanks! 
Could you please share your experience, what's the largest amount of new speaker data, that still achieved the improvement with the mixed setup over the direct fine-tuning?

**Reply by gedefet** [reply]
I have so far 45min of pure new speaker data. At this point always more data -> more improvement. I'm continuing adding more data

**Reply by arinaruck** [reply]
More new speaker data, correct?
And this is with the total amount of data fixed at 5h? Have you tried just the direct fine-tuning with the same amount of new speaker data(45 mins)?

**Reply by gedefet** [reply]
Yes. It works better mixing the data. I'm now training with 2h of LJSpeech+5h of my speaker

**Reply by arinaruck** [reply]
Cool, thanks so much for the update!
Btw, I found [this paper](https://arxiv.org/pdf/2110.05798.pdf) on the same topic, which apparently inspired the mixed training block in the tutorial. 
Maybe, it could be useful to you, if you haven't seen it yet

**Reply by gedefet** [reply]
Hey that's great. Have you found that in google or do you look in some place in specific?
Thanks!

**Reply by arinaruck** [reply]
Found it while reading other threads on fine-tuning fastpitch

**Reply by arinaruck** [reply]
Hey, one more question about the mixed training. 
So you said that the mixed data works better for you, how are you comparing the experiments? 
Is the difference measured by a binary side-by-side comparison, some kind of perception metric like MOS or is the difference so significant that you can hear it just by comparing a few samples yourself?

**Reply by gedefet** [reply]
The difference in all of these variations is noticeable by hearing the generated samples. I'm not using a formal comparison method at the moment. Also, I've trained HiFiGAN only on the new speaker, and not with the full dataset but only on the very high-quality samples. I'm using Spotify podcasts.

I have to try also putting only new samples in the validation dataset of the training dataset. The balance between train and val also remains an unknown. I'm using about 2-3% of the samples for validation. 

Apart from these, at the moment I couldn't make that validation loss decreases over time in finetuning fastpitch. It always increases, despite the samples may turn better. This does not happen with HiFiGAN, as the val loss always decreases.

I'd love to hear about your experience with this.

**Reply by arinaruck** [reply]
Yes, I'm also experiencing the plato / increase of the validation loss, when training fastpitch
I think this is not a problem, taking into the account the one to many nature of the TTS task. There are multiple possible pronunciations for the query text, therefore the resulting audio might sound natural, but not the same as the reference (hence no decrease in the validation loss)
I've experienced similar behaviour with the Tacotron2 model as well

**Reply by gedefet** [reply]
Thank you, that makes sense.

Also, any idea/experience about the balance between train/val data?

**Reply by arinaruck** [reply]
In my experience selecting the hyperparameters or doing early stopping based on the validation set metrics is unpractical, due to the one-to-many quality we discussed before. 
Therefore it seems that the validation set is useful for monitoring the model's performance in inference mode on different stages of the training. So my go to is just a small fixed validation set for sanity checking and as much as possible of the data in the training set

**Comment by gedefet** [comment]
@arinaruck , if you have gone with mixed finetuning, in the small fixed validation dataset approach you mentioned, do/did you use samples from both speakers or only from the new one?

Tkx,
